### PR TITLE
Basic implementation of upload buttons with FormData

### PR DESCRIPTION
**Pull request by maettu**
With this, we can avoid rendering an invisible form in an invisible iframe.
(We still have to render an invisible `<input>`.)

(haha, this is an invisible joke. edit to see it).

### CODE REVIEW

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js [hunk 1]
```diff
@@ -246,18 +243,65 @@ qx.Class.define("callbackery.ui.plugin.Action", {
...
+            file.addListener('changeFileName', function(e){
+                let formData = new FormData();
+                formData.append("file", e.getData());
+                formData.append("name", "onboardingBulkUpload");
+                formData.append("key", "upload_bulk_invite");
+
+                callbackery.data.Server.getInstance().callAsyncSmart(function(cookie){
+                    formData.append('xsc',cookie);
+
+                    var xhr = new XMLHttpRequest();
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js:264]: please use qooxdoo implementation http://www.qooxdoo.org/apps/apiviewer/#qx.io.request.Xhr

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadToolbarButton.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  /**
+   * @param fieldName {String} upload field name
+   * @param label {String} button label
+   * @param icon {String} icon path
+   * @param command {Command} command instance to connect with
+   */
+
+  construct: function(fieldName, label, icon, command)
+  {
+    this.base(arguments, fieldName, label, icon, command);
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadToolbarButton.js:25]: UploadButton does not support fieldName

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js (hunk 1 above)
Comment (maettu) [reply lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js:264]: I don't immediately get this to work. I don't know how to make the upload file appear in the correct location in the request.

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadToolbarButton.js (hunk 2 above)
Comment (maettu) [reply lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadToolbarButton.js]: removed

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js [hunk 3]
```diff
@@ -246,18 +243,63 @@ qx.Class.define("callbackery.ui.plugin.Action", {
...
+                    }, this);
+                    req.addListener("error", function(e) {
+                        callbackery.ui.MsgBox.getInstance().error(
+                            that.tr("Network error"),
+                            that.tr("Upload of file failed")
+                        );
+                        return;
+                    }, this);
+
+                    req.send(formData);
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js:298]: das ist falsch .. send nimmt kein argument
```suggestion
                    req.setRequestData(formData);
                    req.send();
```

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js [hunk 4]
```diff
@@ -246,18 +243,63 @@ qx.Class.define("callbackery.ui.plugin.Action", {
...
+
+                callbackery.data.Server.getInstance().callAsyncSmart(function(cookie){
+                    formData.append('xsc',cookie);
+                    let req = new qx.io.request.Xhr("/upload", "POST");
+                    req.addListener("success", function(e) {
+                        let req = e.getTarget();
+                        file.clear(); // So we can upload the same file multiple times in a row.
+                        let response = {};
+                        try {
+                            response = qx.lang.Json.parse(req.getResponse());
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Action.js:269]: das ist odd ... ich haette erwartet das getResponse automatisch parst

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * Adapted from the uploadwidget extension.
+ */
+qx.Class.define("callbackery.ui.form.UploadButton", {
+  extend : qx.ui.form.Button,
+
+  construct: function(label, icon, command) {
+    this.base(arguments, label, icon, command);
+
+    this.__inputEl = this._createInput();
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js:12]: damit fuerhst du das 2 mal aus

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+  events: {
+    changeFileName: 'qx.event.type.Data'
+  },
+
+  properties: {
+  },
+
+  members: {
+    _createInput: function() {
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js:28]:
```suggestion
   __inputEl: null,
    _createInput: function() {
```

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            // a huge button that covers
+            // the area of the upload button
+            fontSize: '400px'
+        };
+        if ( qx.core.Environment.get('browser.name') == 'ie' ) {
+            css.height = '100%';
+            css.width = '200%';
+        }
+
+        control =  new qx.html.Element('input',css, {
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js:60]:
```suggestion
        control =  this.__inputEl = new qx.html.Element('input',css, {
```

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js (hunk 5 above)
Comment (maettu) [reply lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/UploadButton.js]: Ich denke, das ist so vom Original kopiert.

**Pull request closed without merging**